File: src/main/java/com/raychenon/leetcode/KClosestPointsToOrigin.java
package com.raychenon.leetcode;

import java.util.Arrays;

/**
 * User: raychenon
 * Date: 31/5/2020
 * https://leetcode.com/problems/k-closest-points-to-origin/
 */
public class KClosestPointsToOrigin {

    /**
     * Time complexity: O(N*log(N)) , due to sorting . N is the number of points
     * Space complexity: O(N)
     * @param points
     * @param K
     * @return
     */
    public int[][] kClosestSort(int[][] points, int K) {
        int N = points.length;
        int[] dists = new int[N];
        for (int i = 0; i < N; ++i)
            dists[i] = distance(points[i]);

        Arrays.sort(dists);
        int distK = dists[K - 1];

        int[][] ans = new int[K][2];
        int t = 0;
        for (int i = 0; i < N; i++) {
            if (distance(points[i]) <= distK) {
                ans[t++] = points[i];
            }
        }
        return ans;
    }

    private int distance(int[] point) {
        return point[0] * point[0] + point[1] * point[1];
    }
}

File: src/main/java/com/raychenon/leetcode/KClosestPointsToOrigin.kt
package com.raychenon.leetcode

/**
 * User: raychenon
 * Date: 30/5/2020
 * https://leetcode.com/problems/k-closest-points-to-origin/
 */
object KClosestPointsToOrigin {

    fun kClosestOriginal(points: Array<IntArray>, K: Int): Array<IntArray> {
        var min = Integer.MAX_VALUE
        val pointsMap = mutableMapOf<IntArray, Int>()

        for (p in points) {
            val dist = distance(p)
            pointsMap.put(p, dist)

            min = Math.min(min, dist)
        }


        val sorted = pointsMap.toList().sortedBy { (_, value) -> value }.toMap()
        var arr = Array(K){ intArrayOf(1,2)}


        var iter = sorted.entries.iterator()
        for(i in 0 until K){
            arr[i] = iter.next().key as IntArray
        }

        return arr
    }

    // distance from (0,0)
    fun distance(point: IntArray): Int {
        return point[0] * point[0] + point[1] * point[1]
    }

    fun List<IntArray>.toIntArray(size: Int): Array<IntArray> {
        val iter = iterator()
        return Array(size) { iter.next() }
    }
}